### PR TITLE
[X86][FixupSetCC] Substitute setcc + zext pair with setzucc if possible

### PR DESCRIPTION
**Pull request by KanRobert**

### CODE REVIEW

File: llvm/test/CodeGen/X86/apx/setzucc.ll [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+define i16 @i8(i8 %x) nounwind {
+; CHECK-LABEL: i8:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    cmpb $3, %dil
+; CHECK-NEXT:    setzuae %al
+; CHECK-NEXT:    # kill: def $ax killed $ax killed $eax
+; CHECK-NEXT:    retq
+  %t0 = icmp ugt i8 %x, 2
+  %if = select i1 %t0, i8 1, i8 0
+  %zext = zext i8 %if to i16
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/apx/setzucc.ll:13]: How about `zext i1 %t0` to i16?

File: llvm/lib/Target/X86/X86FixupSetCC.cpp [hunk 2]
```diff
@@ -110,17 +117,23 @@ bool X86FixupSetCCPass::runOnMachineFunction(MachineFunction &MF) {
...
+      // inserting the setcc/setzucc result into the low byte of the zeroed
+      // register.
       Register ZeroReg = MRI->createVirtualRegister(RC);
-      BuildMI(MBB, FlagsDefMI, MI.getDebugLoc(), TII->get(X86::MOV32r0),
-              ZeroReg);
+      if (ST->hasZU()) {
+        MI.setDesc(TII->get(X86::SETZUCCr));
+        BuildMI(*ZExt->getParent(), ZExt, ZExt->getDebugLoc(),
+                TII->get(TargetOpcode::IMPLICIT_DEF), ZeroReg);
+      } else
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86FixupSetCC.cpp:128]: Add parentheses for else block.

File: llvm/lib/Target/X86/X86FixupSetCC.cpp [hunk 3]
```diff
@@ -110,17 +117,23 @@ bool X86FixupSetCCPass::runOnMachineFunction(MachineFunction &MF) {
...
+        // Initialize a register with 0. This must go before the eflags def
+        BuildMI(MBB, FlagsDefMI, MI.getDebugLoc(), TII->get(X86::MOV32r0),
```
Comment (goldsteinn) [inline llvm/lib/Target/X86/X86FixupSetCC.cpp:130]: Was MBB incorrect here?

File: llvm/lib/Target/X86/X86FixupSetCC.cpp [hunk 4]
```diff
@@ -98,9 +106,8 @@ bool X86FixupSetCCPass::runOnMachineFunction(MachineFunction &MF) {
         continue;
 
       // On 32-bit, we need to be careful to force an ABCD register.
-      const TargetRegisterClass *RC = MF.getSubtarget<X86Subtarget>().is64Bit()
-                                          ? &X86::GR32RegClass
-                                          : &X86::GR32_ABCDRegClass;
+      const TargetRegisterClass *RC =
```
Comment (goldsteinn) [inline llvm/lib/Target/X86/X86FixupSetCC.cpp:109]: Seems unrelated.

Comment (KanRobert) [reply llvm/lib/Target/X86/X86FixupSetCC.cpp:109]: It's kind of related.  There are 2 uses of `X86Subtarget` after this PR, so I define the variable `ST` for it.  Unless we'd like to define the variable even for 1 use,  I can not extract this into a separate NFC change.

File: llvm/lib/Target/X86/X86FixupSetCC.cpp (hunk 3 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86FixupSetCC.cpp:130]: It's correct. `MOV32r0` is in the same bb as `FlagsDefineMI` instead of `INSERT_SUBREG`.

File: llvm/test/CodeGen/X86/apx/setzucc.ll (hunk 1 above)
Comment (KanRobert) [reply llvm/test/CodeGen/X86/apx/setzucc.ll]: Done

File: llvm/lib/Target/X86/X86FixupSetCC.cpp (hunk 2 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86FixupSetCC.cpp]: Done

**Review by phoebewang — APPROVED** [review]
LGTM.
